java/ql/src/utils/GenerateFlowTestCase.qll

```diff
@@ -0,0 +1,555 @@
+    )
+  }
+
+  /**
+   * Returns an inline test expectation appropriate to this CSV row.
+   */
+  string getExpectation() {
+    kind = "value" and result = "// $hasValueFlow"
+    or
+    kind = "taint" and result = "// $hasTaintFlow"
+  }
+
+  /**
+   * Returns a declaration and initialisation of a variable named `instance` if required; otherwise returns an empty string.
+   */
+  string getInstancePrefix() {
+    if
+      callable instanceof Method and
+      not callable.(Method).isStatic() and
+      baseOutput != SummaryComponentStack::argument(-1) and
+      baseInput != SummaryComponentStack::argument(-1)
+    then
+      // In this case `out` is the instance.
+      result = getShortNameIfPossible(callable.getDeclaringType()) + " instance = null;\n\t\t\t"
+    else result = ""
+  }
+
+  /**
+   * Returns the type of the output for this test.
+   */
+  Type getOutputType() {
+    if baseOutput = SummaryComponentStack::return()
+    then result = callable.getReturnType()
+    else
+      exists(int i |
+        baseOutput = SummaryComponentStack::argument(i) and
+        result = getParameterType(callable, i)
+      )
+  }
+
+  /**
+   * Returns the type of the input for this test.
+   */
+  Type getInputType() {
+    exists(int i |
+      baseInput = SummaryComponentStack::argument(i) and
+      result = getParameterType(callable, i)
+    )
+  }
+
+  /**
+   * Returns the Java name for the type of the input to this test.
+   */
+  string getInputTypeString() { result = getShortNameIfPossible(this.getInputType()) }
+
+  /**
+   * Returns the next-highest stack item in `input`, treating the list as circular, so
+   * the top item's successor is `baseInput`, the bottom of the stack.
+   */
+  SummaryComponentStack nextInput(SummaryComponentStack prev) {
+    exists(SummaryComponentStack next | next.tail() = prev and next = input.drop(_) | result = next)
+    or
+    not exists(SummaryComponentStack next | next.tail() = prev and next = input.drop(_)) and
+    result = baseInput
+  }
+
+  /**
+   * Returns a call to `source()` wrapped in `newWith` methods as needed according to `input`.
+   * For example, if the input specification is `ArrayElement of MapValue of Argument[0]`, this
+   * will return `newWithMapValue(newWithArrayElement(source()))`.
+   *
+   * This requires a slightly awkward walk, out from the element above the root (`Argument[0]` above)
+   * climbing up towards the outer `ArrayElement of...`. This is implemented by treating the stack as
+   * a circular list, walking it backwards and treating the root as a sentinel indicating we should
+   * emit `source()`.
+   */
+  string getInput(SummaryComponentStack componentStack) {
+    componentStack = input.drop(_) and
+    (
+      if componentStack = baseInput
+      then result = "source()"
+      else
+        result =
+          "newWith" + contentToken(getContent(componentStack.head())) + "(" +
+            this.getInput(nextInput(componentStack)) + ")"
+    )
+  }
```

Comment:
How about simplifying this as follows:
```
string getInput(SummaryComponentStack stack) {
  stack = input and result = "source()"
  or
  exists(SummaryComponentStack s |
    result = "newWith" + contentToken(getContent(s.head())) + "(" + this.getInput(s) + ")" and
    stack = s.tail()
  )
}
```
Then the desired result will be `this.getInput(baseInput)`. Then you can also delete the confusing `nextInput` above.